libenv/sysinfo.c

```diff
@@ -972,7 +972,7 @@ static void Get3Environment(EvalContext *ctx)
 
         if (*context == '@')
         {
-            if (sscanf(context + 1, "%[^=]=%[^\n]", name, value) == 2)
+            if (sscanf(context + 1, "%255[^=]=%255[^\n]", name, value) == 2)
```

Comment:
As discussed, add these lines above:

```
assert((sizeof(name) - 1) == 255);  // TODO: static_assert()
assert((sizeof(value) - 1) == 255); // TODO: static_assert()
```

It's just an attempt to prevent developers from shooting themselves in the foot.
What if someone copy pastes this code to another function with different size `name`?
What if someone changes `CF_MAXVARSIZE`?
What if someone changes `name` to be a pointer?

Reply:
I added asserts now :+1: Please look at the rework :smiley: 

However I sill think that it is more sensible to set these buffers to the required size (in this example, 256) instead of the macros. That will allow us to remove the `assert` calls and will also optimize the stack memory usage (at the moment we allocate arrays of 4096 bytes to use only 255 max)

Reply:
I agree, except for this part:

> That will allow us to remove the assert calls

We don't want to remove them, see reasons above.

Reply:
> I agree, except for this part:
> 
> > That will allow us to remove the assert calls
> 
> We don't want to remove them, see reasons above.

The final code has buffers of `CF_MAXVARSIZE` size and the `assert` calls, so this discussions is not relevant for the PR any more.

However, what I was suggesting was something like:
```
char strmajor[256], strminor[256];
...
sscanf(strversion, "VERSION = %255s", strmajor);
sscanf(strpatch, "PATCHLEVEL = %255s", strminor);
```

And for that case, I argue that we don't need asserts: you are hardcoding the size of an array locally, just before using it on a `sscanf` also with a hardcoded max size.

Reply:
@lluiscampos I understand, and I argued that you should still have the asserts there, like so:

```
char strmajor[256];
...
assert(sizeof(strmajor) == 256);
sscanf(strversion, "VERSION = %255s", strmajor);
```

The reason is that someone might try to change the 256 in the top of the function, and if the function is big and hard to read, they might miss the 255 in the format string (you can't even search for 256, you would miss it). Similarly, someone could copy the sscanf line, or just the format string, to another function, where the buffer size is different.

C is an unsafe language, code becomes more complicated, less maintainable, and more "copy-pasted" over time. I am just trying to be as safe and explicit as possible. When you write that format string, you are assuming that the buffer is 256 bytes, or more. So, make that assumption explicit, add an assert as close as possible to the code that warrants it.

The `...` in your example could be hundreds of lines of code in the future. You can't really control developers working on the project in the future, you can't guarantee that they don't make mistakes when they try to change it, the best we can do is try to prevent them from making those mistakes, by writing clear/readable code, useful comments and asserts / errors that show when something is wrong.